optimade/server/schemas.py

```diff
@@ -0,0 +1,118 @@
+                properties.update(
+                    retrieve_queryable_properties(
+                        sub_schema, sub_queryable_properties, entry_provider_fields
+                    )
+                )
+            else:
+                properties[name] = {"description": value.get("description", "")}
+                if "unit" in value:
+                    properties[name]["unit"] = value["unit"]
+                # All properties are sortable with the MongoDB backend.
+                # While the result for sorting lists may not be as expected, they are still sorted.
+                properties[name]["sortable"] = value.get("sortable", True)
+
+            if name in entry_provider_fields:
+                # Rename fields in schema according to provider field list
+                properties[f"_{CONFIG.provider.prefix}_{name}"] = properties.pop(name)
+
+    return properties
+
+
+class EntrySchemas:
+    """ Wrapper class for EntryResource schemas used in this server. Any
+    references in the schema are resolved to allow for easy use. If the
+    underyling model is updated, the schema will be refreshed.
+
+    """
+
+    def __init__(self):
+        self._structure_model = StructureResource
+        self._reference_model = ReferenceResource
+        self._links_model = LinksResource
+        self._structures = None
+        self._references = None
+        self._links = None
+
+    def keys(self):
+        return {"structures", "references"}
+
+    def get(self, key: str):
+        return getattr(self, key, {})
+
+    @property
+    def structures(self):
+        if self._structures is None:
+            schema = self.structure_model.schema()
+            entry_provider_fields = CONFIG.provider_fields.get("structures")
+            self._structures = retrieve_queryable_properties(
+                schema,
+                schema["properties"].keys(),
+                entry_provider_fields=entry_provider_fields,
+            )
+
+        return self._structures
+
+    @property
+    def references(self):
+        if self._references is None:
+            schema = self.reference_model.schema()
+            entry_provider_fields = CONFIG.provider_fields.get("references")
+            self._references = retrieve_queryable_properties(
+                schema,
+                schema["properties"].keys(),
+                entry_provider_fields=entry_provider_fields,
+            )
+
+        return self._references
+
+    @property
+    def structure_model(self):
+        return self._structure_model
+
+    @structure_model.setter
+    def structure_model(self, model: EntryResource):
+        self._structures = None
+        self._structure_model = model
+
+    @property
+    def reference_model(self):
+        return self._reference_model
+
+    @reference_model.setter
+    def reference_model(self, model: EntryResource):
+        self._references = None
+        self._reference_model = model
```

Comment:
This looks like code that could be further abstracted to me.
Since there are "only two" repititions essentially, for structures and references, respectively, it can be argued that abstraction is overkill, but it would future proof for calculations and other entry resources.
Have you considered doing something similar to the Adapter classes concerning how to retrieve attributes?